### DISCUSSION TITLE
Support for USB CDC-FTDI on STM32

### DISCUSSION THREAD

**Discussion by medchedli** [hathach/tinyusb#725] [Q&A] [answered]
I am working on a project where we want to replace a computer (the Host) with an STM32 microcontroller. The device uses a standard FTDI USB to serial converter as the main mechanism to communicate between the computer and controller. Instead of using the computer, I want to use STM32F4 as a host to communicate with the device.

STM32 -> USB -> FTDI -> Machine controller

STM32 natively supports USB CDC communication but does not support communication with FTDI chips. On ST website there is an article about an ST partner who developed a commercial USB Stack that supports CDC-FTDI, but since the project is still in an experimental phase I am looking for other alternatives https://www.st.com/en/embedded-software/hcc-usb-stack.html

I would appreciate your expert opinion on this. Does tinyUSB lib have support for communication with FTDI chips ? If yes how to get started on that ?

**Comment by hathach** [comment]
yeah, I used to have a plan to support host FTDI driver as well. I just don't have enough time to implement it. Also, the host stack doesn't support stm32 just yet. It is still WIP, and need lots of development time. My main mcu for developing host stack is lpc18xx (EHCI), once I got all the USBH/HCD API settled. I will look into porting more hcd.

**Comment by medchedli** [comment]
Thanks @hathach 🙏 

Just for future reference maybe this could be helpful. @dkonigsberg shared with us his attempt to write an FTDI driver for STM32. We tried it and the results are promising : https://github.com/dkonigsberg/printalyzer/blob/master/software/firmware/external/usb/class/src/usbh_serial_ftdi.c

**Reply by hathach** [reply]
thanks for the link, all the reference code will help later development.

**Comment by IngHK** [comment]
See #2488.
Any kind of feedback, review and test is welcome.

**Comment by dkonigsberg** [comment]
Just to chime in...  The big blocker for myself using TinyUSB host-side (and likely for the OP as well) is simply the lack of STM32 host support itself.  Without that, its pretty much impossible to test the class drivers.
Right now, the only non-commercial embedded USB stacks with STM32 host support are ST's own middleware USB host library and CherryUSB.  Of course even CherryUSB's own support is limited, as they only support USB peripherals with DMA (which limits it to higher end MCUs with an HS peripheral block).

**Comment by rjackson159** [comment]
If anyone is still interested in this, I was able to get this solution by Renesas to work, it requires very minimal changes to the USBX stack.
https://en-support.renesas.com/knowledgeBase/20627979 
I have working FTDI communication on STM32G0 part using USBX Host stack with the custom read/write functions defined in the Renesas post.

**Comment by HiFiPhile** [comment, marked as answer]
FYI [FTDI CDC host driver](https://github.com/hathach/tinyusb/blob/38842491bb55260c0cc7d7709abe5b0e73a0de61/src/class/cdc/cdc_host.c#L38) is included, HCD (dwc2) for higher-tier STM32 is supported and for lower-tier (fsdev) is pending in #3364
